### PR TITLE
ES6 constructor returns class instead of object upon constructorCache hit.

### PR DESCRIPTION
**Pull request by suwc**
ES6 constructor returns class instead of object upon constructorCache hit.
constructorCache->ctorHasNoExplicitReturnValue set to 'true' for ES6 class
constructors, therefore opcode 'GetNewScObject' is optimized away in
inliner. Fix by de-asserting Flags_HasNoExplicitReturnValue for ES6 class
constructors.

### CODE REVIEW

**Comment by suwc** [comment]
https://github.com/Microsoft/ChakraCore/issues/1496

**Comment by suwc** [comment]
@Microsoft/chakra-es @eebbing @Krovatkin could you take a look pls?

File: lib/Backend/GlobOptFields.cpp [hunk 1]
```diff
@@ -2700,7 +2700,7 @@ GlobOpt::OptNewScObject(IR::Instr** instrPtr, Value* srcVal)
         instr->m_func->GetConstructorCache(static_cast<Js::ProfileId>(instr->AsProfiledInstr()->u.profileId)) : nullptr;
 
     Assert(ctorCache == nullptr || srcVal->GetValueInfo()->IsVarConstant() && Js::JavascriptFunction::Is(srcVal->GetValueInfo()->AsVarConstant()->VarValue()));
-    Assert(ctorCache == nullptr || !ctorCache->typeIsFinal || ctorCache->ctorHasNoExplicitReturnValue);
+    Assert(ctorCache == nullptr || !ctorCache->typeIsFinal || ctorCache->ctorHasNoExplicitReturnValue || isCtorInlined);
```
Comment (rajatd) [inline lib/Backend/GlobOptFields.cpp]: Shouldn't this be a check for a class constructor? We emit NewScObjectNoCtor in cases other than class ctor too.

File: lib/Backend/Lower.cpp [hunk 2]
```diff
@@ -4577,7 +4577,7 @@ bool Lowerer::TryLowerNewScObjectWithFixedCtorCache(IR::Instr* newObjInstr, IR::
         ctorCache = newObjInstr->m_func->GetConstructorCache(static_cast<Js::ProfileId>(newObjInstr->AsProfiledInstr()->u.profileId));
         Assert(ctorCache != nullptr);
         Assert(!ctorCache->skipNewScObject);
-        Assert(!ctorCache->typeIsFinal || ctorCache->ctorHasNoExplicitReturnValue);
+        Assert(!ctorCache->typeIsFinal || ctorCache->ctorHasNoExplicitReturnValue || newObjInstr->m_opcode == Js::OpCode::NewScObjectNoCtor);
```
Comment (rajatd) [inline lib/Backend/Lower.cpp]: >newObjInstr->m_opcode == Js::OpCode::NewScObjectNoCtor) [](start = 85, length = 55)

same as in GloboptFields.cpp

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp [hunk 3]
```diff
@@ -2727,6 +2727,11 @@ void ByteCodeGenerator::EmitFunctionBody(FuncInfo *funcInfo)
     }
     Assert(!pnode->CapturesSyms());
     EmitTopLevelStatement(pnode, funcInfo, false);
+
+    if (funcInfo->IsClassConstructor())
+    {
+        funcInfo->GetParsedFunctionBody()->SetHasNoExplicitReturnValue(false);
```
Comment (ianwjhalliday) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: Is this the right place to set this flag?  Seems like this is something that could be set earlier when the function body object is created.

Comment (suwc) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: Possibly. Note other calls to SetHasNoExplicitReturnValue() won't happen till inside EmitTopLevelStatement(). I think it's more consistent to set the flag here.

---
In reply to: [78957170](https://github.com/Microsoft/ChakraCore/pull/1577#discussion_r78957170) [](ancestors = 78957170)

**Comment by suwc** [comment]
Any other comments or questions?

File: test/es6/rlexe.xml [hunk 4]
```diff
@@ -1331,4 +1331,12 @@
     <compile-flags>-force:deferparse -args summary -endargs</compile-flags>
   </default>
 </test>
+<test>
+  <default>
+    <files>bug_issue_1496.js</files>
+    <compile-flags>-mic:1 -maxsimplejitruncount:2</compile-flags>
```
Comment (tcare) [inline test/es6/rlexe.xml]: Why does dynapogo not work here? Because of the simplejit runcount?

Comment (suwc) [inline test/es6/rlexe.xml]: No, dynapogo works for the test case, but it didn't repro the bug in the first place. And given the flags running dynapogo here has little value.

Comment (tcare) [inline test/es6/rlexe.xml]: Looks like other tests that include -mic:1 allow dynapogo. I don't see the harm in having it.

**Comment by tcare** [comment]
Looks good

File: test/es6/rlexe.xml (hunk 4 above)
Comment (suwc) [inline test/es6/rlexe.xml]: No harm - it's value vs. (a little extra) run time.
Let's get 2nd opinion.
@ianwjhalliday, what do you think?

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp (hunk 3 above)
Comment (pleath) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: I'm guessing that other cases set the flag when they visit the return statement node. I think for class ctors we should either do as Ian suggests and set the flag in StartBindFunction, which is a friendlier place, or have the parser emit an explicit return node.

File: test/es6/rlexe.xml (hunk 4 above)
Comment (pleath) [inline test/es6/rlexe.xml]: Let's have dynapogo run the test. More aggressive jitting in such a case can expose issues with inlining, etc., down the road.

**Pull request merged**